### PR TITLE
optimize file write performance

### PR DESCRIPTION
**Pull request by CoffeeLatte007**
### Ⅰ. Describe what this PR did
1. optimize file write performance and optimize saveHistory
2. By the way，optimize closes gracefully.
3. add sync write mode，it is suitable for data consistency scenarios.

mac:  2.2 GHz Intel Core i7.   4 core  .
per_thread_write_file = per_thread_trx_num * 9
1. new vs old
===========new file write aync mode====================
thread nums:1, per_thread_trx_num:65535 ,cost2600
thread nums:4, per_thread_trx_num:65535 ,cost13329
thread nums:8, per_thread_trx_num:65535 ,cost27613
thread nums:16, per_thread_trx_num:65535 ,cost57764
===========old file write aync mode====================
thread nums:1, per_thread_trx_num:65535 ,cost17816
thread nums:4, per_thread_trx_num:65535 ,cost42218
thread nums:8, per_thread_trx_num:65535 ,cost87016
thread nums:16, per_thread_trx_num:65535 ,cost176943
Performance improved by more than 60%
2. In new write mode, new saveHistory vs old saveHistory
set trx_time_out = 1000ms,  each saveHistory will save 3000 data.
===========new saveHistory ====================
thread nums:16, per_thread_trx_num:65535 ,cost 60559
===========old saveHistory ====================
thread nums:16, per_thread_trx_num:65535 ,cost90189

### Ⅱ. Does this pull request fix one issue?

https://github.com/seata/seata/issues/858

### CODE REVIEW

File: server/src/main/resources/nacos-config.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ store.file.dir=file_store/data
 store.file.max-branch-session-size=16384
 store.file.max-global-session-size=512
 store.file.file-write-buffer-cache-size=16384
+store.file.flush-disk-mode=async
 store.file.session.reload.read_size=100
 store.db.driver_class=com.mysql.jdbc.Driver
 store.db.url=jdbc:mysql://localhost:3306/fescar_demo
```
Comment (xingfudeshi) [inline server/src/main/resources/nacos-config.txt]: seata_demo

File: server/src/test/java/WriteStoreMultithreadTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+/**
+ * The interface Session storable.
+ *
+ * @author lizhao
+ */
+public class WriteStoreMultithreadTest {
+    private static String vgroup = "vgroupMock";
+    private static String appname = "appnameMock";
+    private static String instname = "fescarMocK";
```
Comment (xingfudeshi) [inline server/src/test/java/WriteStoreMultithreadTest.java]: seataMock

File: server/src/test/java/WriteStoreMultithreadTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    private static int per_thread_trx_num = 65535;
+    private static int threadNum = 16;
+    private static CountDownLatch countDownLatch = new CountDownLatch(threadNum);
+    public static void main(String[] args) throws Exception {
```
Comment (xingfudeshi) [inline server/src/test/java/WriteStoreMultithreadTest.java]: please use @Test instead of main method,and some Asserts to verify it.

Comment (CoffeeLatte007) [reply server/src/test/java/WriteStoreMultithreadTest.java]: Reference WriteStoreTest, it is not a unit test

File: server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java [hunk 4]
```diff
@@ -121,13 +161,94 @@ private void initFile(String fullFileName) throws IOException {
...
-        TransactionWriteFuture writeFutureRequest = new TransactionWriteFuture(session, logOperation);
+        writeSessionLock.lock();
+        long curFileTrxNum;
         try {
-            if (transactionWriteFutureQueue.offer(writeFutureRequest, MAX_ENQUEUE_MILLS, TimeUnit.MILLISECONDS)) {
-                return writeFutureRequest.get();
+            if (!writeDataFile(new TransactionWriteStore(session, logOperation).encode())) {
+                return false;
+            }
+            lastModifiedTime = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (CoffeeLatte007) [reply server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java]: why use file. lastModifiedTime? I'm not sure it get the lastModifiedTime from disk every time

File: server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java [hunk 5]
```diff
@@ -260,132 +392,251 @@ private void closeFile(RandomAccessFile raf) {
...
+        }
+
+
+        public void wakeupCustomer() {
+            this.countDownLatch.countDown();
+        }
+
+        public void waitForFlush(long timeout) {
+            try {
+                this.countDownLatch.await(timeout, TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (CoffeeLatte007) [reply server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java]: i think it should not check return value as rocketmq. even it happend time out , we should not block our process

File: server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java [hunk 6]
```diff
@@ -260,132 +392,251 @@ private void closeFile(RandomAccessFile raf) {
...
                     }
+                    this.requestsRead.clear();
+                } else {
+                    flushOnCondition(currFileChannel);
                 }
             }
+        }
 
+        private void closeAndFlush(CloseFileRequest req) {
+            long diff = FILE_TRX_NUM.get() - FILE_FLUSH_NUM.get();
```
[the remark this answers is not in the document]
Comment (CoffeeLatte007) [reply server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java]: syncFlush will weakUp. Close  does not need

File: server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java [hunk 7]
```diff
@@ -260,132 +392,251 @@ private void closeFile(RandomAccessFile raf) {
...
+                return true;
+            } catch (IOException exx) {
+                LOGGER.error("write data file error:" + exx.getMessage());
+            }
+        }
+        LOGGER.error("write dataFile failed,retry more than :" + MAX_WRITE_RETRY);
+        return false;
+    }
+
+    interface StoreRequest {
```
[the remark this answers is not in the document]
Comment (CoffeeLatte007) [reply server/src/main/java/io/seata/server/store/FileTransactionStoreManager.java]: fix it

**Pull request merged**